### PR TITLE
Issue #466 Add NMatrix#magic function

### PR DESCRIPTION
**Pull request by shahsaurabh0605**
Implemented the magic function as described in the issue [#466](https://github.com/SciRuby/nmatrix/issues/466).

### CODE REVIEW

**Comment by shahsaurabh0605** [comment]
Sorry for such gawky variable naming and other silly things. Hurry to send the pull request led to those. Let me know if any other changes have to be made.

File: lib/nmatrix/shortcuts.rb [hunk 1]
```diff
@@ -465,7 +465,137 @@ def random(shape, opts={})
...
+
+      temp=nm[k,0] 
+      nm[k,0]=nm[k+half_shape,0] 
+      nm[k+half_shape,0]=temp
 
+      temp=nm[k+half_shape,k]
+      nm[k+half_shape,k]=nm[k,k] 
+      nm[k,k]=temp
+    end   
+    private :odd_magic, :doubly_even_magic, :singly_even_magic
```
Comment (v0dro) [inline lib/nmatrix/shortcuts.rb]: @mohawkjohn shouldn't these functions be placed in a separate helper module?

Comment (mohawkjohn) [inline lib/nmatrix/shortcuts.rb]: Like, just in a separate file? Can you give an example of what you mean, @v0dro?

Comment (v0dro) [inline lib/nmatrix/shortcuts.rb]: So you can create a module `MagicHelpers` inside this same file and add the functions listed as private to the module as class methods.

So:
``` ruby
class NMatrix
  module MagicHelpers
    class << self
      # methods
    end
  end
end

def self.magic n
  # call as MagicHelpers.method
end
```

@mohawkjohn this is what you had told me to do when I was writing a function for solving tridiagonal equations. That PR got old so I closed it: https://github.com/SciRuby/nmatrix/pull/301/files

Comment (mohawkjohn) [inline lib/nmatrix/shortcuts.rb]: Ah, yes, that would be the best way to do it.

**Comment by shahsaurabh0605** [comment]
I have worked my way to implement the private methods in the module. Please have a look if it is implemented correctly.

**Comment by v0dro** [comment]
looks good to me. why are tests failing?

**Comment by shahsaurabh0605** [comment]
My previous two commits passed all the tests. I am also not getting why this time the tests are failing.

**Comment by v0dro** [comment]
it exceeded the max time limit haha. just restart the failing tests.

**Comment by shahsaurabh0605** [comment]
Actually, I am new to travis ci checks. What does restart the failing tests mean? Does it mean that I should submit my commit again?

**Comment by v0dro** [comment]
No. Just google restarting Travis builds.

**Comment by shahsaurabh0605** [comment]
I found that that the solution would be to close the PR and open it again to trigger the build. Is there any alternate solution possible? 
Sorry for the trouble.

**Comment by mohawkjohn** [comment]
I've restarted it.

**Comment by shahsaurabh0605** [comment]
Well, time limit exceeded once again!!

**Comment by shahsaurabh0605** [comment]
What do you mean by other dtypes? magic function takes shape as an argument and returns a matrix which is constructed of integers from 1 to n^n.

File: spec/shortcuts_spec.rb [hunk 2]
```diff
@@ -159,6 +159,38 @@
     end
   end
   
+  context "::magic" do
+    it "creates a matrix with numbers from 1 to n^n(n squared)" do
```
Comment (Arafatk) [inline spec/shortcuts_spec.rb]: @shahsaurabh0605 I believe the other data types are  `:int16,:int32,:int64` .  So Adding something similar to   
`data_types=[:int8,:int16,:int32,:int64]`
    `data_types.each do |dtype|` will help you run the spec tests for all data types.

Comment (v0dro) [inline spec/shortcuts_spec.rb]: There is a constant called `ALL_DTYPES` in `spec_helper.rb`.

File: spec/shortcuts_spec.rb [hunk 3]
```diff
@@ -159,6 +159,39 @@
     end
   end
   
+  context "::magic" do
+  
+    [:int64,:int32,:int16,:int8].each do |dtype|
+      context dtype do  
+        it "creates a matrix with numbers from 1 to n^n(n squared)" do
+          a = NMatrix.magic(3)
```
Comment (shahsaurabh0605) [inline spec/shortcuts_spec.rb]: I have made the required changes. 
@v0dro Is there any need to check for all the dtypes as we are working with only integers over here?

Comment (v0dro) [inline spec/shortcuts_spec.rb]: No need. How about adding a `INTEGER_DTYPES` constant to `spec_helper.rb` and using that here instead of the Array?

Comment (v0dro) [inline spec/shortcuts_spec.rb]: Why are you checking all dtypes in the first place? You've not even used the `dtype` variable anywhere in your code.

Comment (v0dro) [inline spec/shortcuts_spec.rb]: Those checks for all integer data types are unnecessary.

File: lib/nmatrix/shortcuts.rb [hunk 4]
```diff
@@ -465,7 +551,56 @@ def random(shape, opts={})
...
+    #    NMatrix.magic(6) # => [  [31,  9,  2, 22, 27, 20]
+    #                             [ 3, 32,  7, 21, 23, 25]
+    #                             [35,  1,  6, 26, 19, 24]
+    #                             [ 4, 36, 29, 13, 18, 11]
+    #                             [30,  5, 34, 12, 14, 16]
+    #                             [ 8, 28, 33, 17, 10, 15] ]
+    #
+    def magic(shape)
+      if shape==2
+        raise(ArgumentError, "shape of two is not allowed")
```
Comment (v0dro) [inline lib/nmatrix/shortcuts.rb]: This `if` can be written in one line.

File: lib/nmatrix/shortcuts.rb [hunk 5]
```diff
@@ -465,7 +551,56 @@ def random(shape, opts={})
...
+      end
+      nm= NMatrix.new([shape,shape],0)
```
Comment (v0dro) [inline lib/nmatrix/shortcuts.rb]: If you're going to populate the matrix piecewise later on (and don't depend on it being initialized to `0`), there's no need to pass that `0` as the second argument. Let it contain garbage initially.

File: lib/nmatrix/shortcuts.rb [hunk 6]
```diff
@@ -465,7 +551,56 @@ def random(shape, opts={})
...
+      if shape%2!=0
+        MagicHelpers.odd_magic(nm,shape)
```
Comment (v0dro) [inline lib/nmatrix/shortcuts.rb]: Not using the parenthesis after a method call is more ruby-like.

File: lib/nmatrix/shortcuts.rb [hunk 7]
```diff
@@ -34,6 +34,92 @@
...
+          end
+        end
+      end
+    
+      def singly_even_magic(nm,shape)
+        half_shape = shape/2
+	    k = (shape-2)/4
+	    swap_col= NMatrix.new([shape], dtype: :int32)
+        index=0 
+        mini_magic= NMatrix.new([half_shape,half_shape],0)
```
Comment (v0dro) [inline lib/nmatrix/shortcuts.rb]: Again, does the algorithm depend on the matrix being initialized to `0`?

File: spec/shortcuts_spec.rb (hunk 3 above)
Comment (shahsaurabh0605) [inline spec/shortcuts_spec.rb]: This seems to be a good idea as we are having the array declared for both ALL_DTYPES and NON_INTEGER_DTYPES.

File: lib/nmatrix/shortcuts.rb [hunk 8]
```diff
@@ -34,6 +34,92 @@
...
+        end
+      end
+    
+      def doubly_even_magic(nm,shape)
+        mini_square_num = shape/4
+        cnt = 1     
+        inv_cnt = shape*shape
+        shape.times do |i|
+          shape.times do |j|
+            if j>=mini_square_num && j<shape-mini_square_num
```
Comment (v0dro) [inline lib/nmatrix/shortcuts.rb]: Dont use `&&`, use `and`. This is not C++.

File: lib/nmatrix/shortcuts.rb [hunk 9]
```diff
@@ -34,6 +34,92 @@
...
+          end
+            nm[row,col]= i
+        end
+      end
+    
+      def doubly_even_magic(nm,shape)
+        mini_square_num = shape/4
+        cnt = 1     
+        inv_cnt = shape*shape
+        shape.times do |i|
```
Comment (v0dro) [inline lib/nmatrix/shortcuts.rb]: Use `row` and `col` instead of `i` and `j`. Lesser abbreviates => Better code.

File: lib/nmatrix/shortcuts.rb [hunk 10]
```diff
@@ -34,6 +34,92 @@
...
 class NMatrix
 
+  # Methods for generating magic matrix.
+  module MagicHelpers
+    class << self
+      def odd_magic(nm,shape)
+        row= shape-1
+        col= shape/2       
+        nm[row,col]= 1
+        (2..shape*shape).each do |i|
```
Comment (v0dro) [inline lib/nmatrix/shortcuts.rb]: What is `i` exactly? The variable name should exactly specify what's it going to do.

File: lib/nmatrix/shortcuts.rb [hunk 11]
```diff
@@ -34,6 +34,92 @@
...
+		      else 
+		        nm[i,j] = inv_cnt
+		      end
+		    elsif i<mini_square_num || i>=shape-mini_square_num
+		      nm[i,j]=cnt
+		    else
+              nm[i,j] = inv_cnt
+            end
+          cnt+=1
+          inv_cnt-=1
```
Comment (v0dro) [inline lib/nmatrix/shortcuts.rb]: Whats `cnt`? Don't use abbreviations.

File: lib/nmatrix/shortcuts.rb [hunk 12]
```diff
@@ -34,6 +34,92 @@
...
+          cnt+=1
+          inv_cnt-=1  
+          end
+        end
+      end
+    
+      def singly_even_magic(nm,shape)
+        half_shape = shape/2
+	    k = (shape-2)/4
+	    swap_col= NMatrix.new([shape], dtype: :int32)
```
Comment (v0dro) [inline lib/nmatrix/shortcuts.rb]: Indentation is wrong here.

File: lib/nmatrix/shortcuts.rb [hunk 13]
```diff
@@ -34,6 +34,92 @@
...
+        index=0 
+        mini_magic= NMatrix.new([half_shape,half_shape],0)
+        odd_magic(mini_magic,half_shape)
+        half_shape.times do |i|
+          half_shape.times do |j|
+            nm[i,j] = mini_magic[i,j]  	
+      	    nm[i+half_shape,j+half_shape] = mini_magic[i,j]+half_shape*half_shape
```
Comment (v0dro) [inline lib/nmatrix/shortcuts.rb]: indentation

File: lib/nmatrix/shortcuts.rb [hunk 14]
```diff
@@ -34,6 +34,92 @@
...
+        end
+      end
+    
+      def singly_even_magic(nm,shape)
+        half_shape = shape/2
+	    k = (shape-2)/4
+	    swap_col= NMatrix.new([shape], dtype: :int32)
+        index=0 
+        mini_magic= NMatrix.new([half_shape,half_shape],0)
+        odd_magic(mini_magic,half_shape)
```
Comment (v0dro) [inline lib/nmatrix/shortcuts.rb]: don't use parentheses in ruby method calls wherever possible.

File: lib/nmatrix/shortcuts.rb (hunk 5 above)
Comment (shahsaurabh0605) [inline lib/nmatrix/shortcuts.rb]: As in the method I am going to check for zero initially, it is necessary to initialize it to zero.

File: lib/nmatrix/shortcuts.rb (hunk 6 above)
Comment (shahsaurabh0605) [inline lib/nmatrix/shortcuts.rb]: Actually in the complete file we are using parenthesis after the method, so I decided to use parenthesis. Should I remove it from all the methods?

File: lib/nmatrix/shortcuts.rb (hunk 7 above)
Comment (shahsaurabh0605) [inline lib/nmatrix/shortcuts.rb]: Yes, here the algorithm depends on the matrix being initialized to 0. Absence of which will give incorrect answers.

File: lib/nmatrix/shortcuts.rb [hunk 15]
```diff
@@ -465,7 +551,54 @@ def random(shape, opts={})
...
+    #                             
+    #    NMatrix.magic(6) # => [  [31,  9,  2, 22, 27, 20]
+    #                             [ 3, 32,  7, 21, 23, 25]
+    #                             [35,  1,  6, 26, 19, 24]
+    #                             [ 4, 36, 29, 13, 18, 11]
+    #                             [30,  5, 34, 12, 14, 16]
+    #                             [ 8, 28, 33, 17, 10, 15] ]
+    #
+    def magic(shape)
+      if shape==2 then raise(ArgumentError, "shape of two is not allowed") end
```
Comment (v0dro) [inline lib/nmatrix/shortcuts.rb]: A better way is `raise(ArgumentError, "shape of two is not allowed") if  shape == 2`

File: spec/shortcuts_spec.rb [hunk 16]
```diff
@@ -159,6 +159,39 @@
     end
   end
   
+  context "::magic" do
+  
+    INTEGER_DTYPES.each do |dtype|
```
Comment (v0dro) [inline spec/shortcuts_spec.rb]: why is this here? Where in the test are you using the variable (dtype)?

File: spec/shortcuts_spec.rb [hunk 17]
```diff
@@ -159,6 +159,39 @@
...
+          c = NMatrix.magic(6)
+          magic6= NMatrix.new([6,6], [31, 9, 2, 22, 27, 20, 3, 32, 7, 21, 23, 25, 35, 1, 6, 26, 19, 24, 4, 36, 29, 13, 18, 11, 30, 5, 34, 12, 14, 16, 8, 28, 33, 17, 10, 15]) 
+          expect(c.stype).to eq(:dense)
+          expect(c).to eq magic6  
+        end  
+      end 
+    end
+    
+    it "shape of two is not allowed" do
+      expect { NMatrix.magic(2) }.to raise_error
```
Comment (v0dro) [inline spec/shortcuts_spec.rb]: `raise_error(ArgumentError)`

File: lib/nmatrix/shortcuts.rb [hunk 18]
```diff
@@ -34,6 +34,92 @@
...
+            end
+            count+=1
+            inv_count-=1  
+          end
+        end
+      end
+    
+      def singly_even_magic(nm,shape)
+        half_shape = shape/2
+        k = (shape-2)/4
```
Comment (v0dro) [inline lib/nmatrix/shortcuts.rb]: what is k? What does it do? Use more verbose variable names. This applies everywhere.

**Comment by shahsaurabh0605** [comment]
Is there anything left to be implemented in the function?

**Comment by shahsaurabh0605** [comment]
I have attempted to make all the changes as suggested.
Should I make an array ALL_STYPES in spec_helper.rb? as this can be used in many of the specs.

File: spec/shortcuts_spec.rb [hunk 19]
```diff
@@ -159,6 +159,40 @@
     end
   end
   
+  context "::magic" do
+  
+    [:dense, :list, :yale].each do |stype|
```
Comment (v0dro) [inline spec/shortcuts_spec.rb]: step 1: Remove this loop and the context stype.

File: spec/shortcuts_spec.rb [hunk 20]
```diff
@@ -159,6 +159,40 @@
...
+      context stype do
+ 	
+        ALL_DTYPES.each do |dtype|
```
Comment (v0dro) [inline spec/shortcuts_spec.rb]: step 2: remove this loop too and context dtype.

File: spec/shortcuts_spec.rb [hunk 21]
```diff
@@ -159,6 +159,40 @@
...
+          context dtype do  
+            it "creates a matrix with numbers from 1 to n^n(n squared)" do
```
Comment (v0dro) [inline spec/shortcuts_spec.rb]: step 3: just keep this it block

File: spec/shortcuts_spec.rb [hunk 22]
```diff
@@ -159,6 +159,40 @@
...
+              a = NMatrix.magic(3)
+              magic3 = NMatrix.new([3,3], [4, 9, 2, 3, 5, 7, 8, 1, 6], dtype: dtype, stype: stype)
```
Comment (v0dro) [inline spec/shortcuts_spec.rb]: The reason why I'm saying so is because there is NOWHERE in the `.magic` function where you specify the dtype or the stype, so _those loops are not necessary_.

**Comment by shahsaurabh0605** [comment]
Done.

File: lib/nmatrix/shortcuts.rb [hunk 23]
```diff
@@ -34,6 +34,92 @@
...
+            count += 1
+            inv_count -= 1  
+          end
+        end
+      end
+    
+      def singly_even_magic(nm, shape)
+        half_shape = shape / 2
+        complementary_pair = (shape - 2) / 4
+        swap_col = NMatrix.new([shape])
```
Comment (shahsaurabh0605) [inline lib/nmatrix/shortcuts.rb]: swap_col requires the new nmatrix to be of integer dtype.

File: lib/nmatrix/shortcuts.rb [hunk 24]
```diff
@@ -34,6 +34,92 @@
...
+        index = 0 
+        mini_magic = NMatrix.new([half_shape,half_shape], 0, dtype: nm.dtype)
```
Comment (shahsaurabh0605) [inline lib/nmatrix/shortcuts.rb]: As mini_magic is passed in odd_magic method, I have assigned its dtype same as that of nm.

File: spec/spec_helper.rb [hunk 25]
```diff
@@ -32,7 +32,7 @@
 
 ALL_DTYPES = [:byte,:int8,:int16,:int32,:int64, :float32,:float64, :object,
   :complex64, :complex128]
-
+
```
Comment (shahsaurabh0605) [inline spec/spec_helper.rb]: I have removed the INTEGER_DTYPES array but I think it can be kept here for future references. Also, if you suggest ALL_STYPES array can also be made for the specs.

**Comment by shahsaurabh0605** [comment]
Are all the changes done correctly?

**Comment by mohawkjohn** [comment]
Looking much better — just the one comment now. =) Good work.

**Comment by shahsaurabh0605** [comment]
I think this should be done.

**Comment by shahsaurabh0605** [comment]
Is there anything left to do in this function?
